Clean up logging

ShadowNinja opened the pull request:
* Remove the ugly C-style debug stream system, replacing it with the
    StreamLogOutput and FileLogOutput classes.  dstream becomes just another
    log level, except that it's always printed.  This adds timestamps and
    thread names to `dstream`'d messages, removing the need for the `DTIME` macro.
   `dstream`'s old functionality is replaced by `rawstream`.
  * Move global logging functions into a Logger class and use a global logger.
  * Add (and use) warning log level.
  * Map Irrlicht log event levels to Minetest log event levels and shorten the
    prefix to "Irrlicht: ".  This adds timestamps, thread names, and filtering
    to the Irrlicht log messages. EDIT: Merged as a93838707a9aab104c20b6a971d8c622184a508f.
  * Don't set the Irrlicht log level to the file log level as this affects
    console output too.  Filter Irrlicht log messages ourselves instead (which we can do because the levels are mapped now).
  * Rename macros with two leading underscores (these names are reserved for
    the compiler/library implementation).
  * Fix some instances where threads didn't de-register themselves from logging.

To do:
  * [x] Fix `setLevelSilenced()` so that it works only on the level passed, instead of all levels (this causes `minetest --run-unittests --info` to not print the `Running Test*` lines in `infostream`, because `LL_ERROR` was disabled, but it affected all log levels).  --  (this was NOT introduced by these changes)
  * [ ] Fix Irrlicht not passing the main menu's "Loaded texture" messages to the event handler as log events, instead printing them directly to `stdout` with `printf`.

est31 commented:
great this fixes #2263

nerzhul commented:
Seems good to me

rubenwardy commented:
+1 for adding warning level, this is needed.

C1ffisme commented:
By the way, could you possibly make logging remove any text that is older than say, 2 years (customizable in minetest.conf)?

Debug.txt takes several **MEGABYTES** just having logs that don't matter anymore.

ShadowNinja commented:
**Updated**: Don't set Irrlicht's log level to the file log level (because it applies to console logging too), filter messages ourselves instead.  Manifests a new bug (although this was already a bug to begin with...), which may be an Irrlicht bug (it's rather hard to debug so I don't really know).  Also makes the engine make more use of `warningstream`.

@C1ffisme: Yes, but I won't in this PR.

proller commented:
Somebody stop him!

Also who delete my comments?

sfan5 commented on src/log.h:
```diff
@@ -20,83 +20,156 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-
+extern std::ostream *derr_server_ptr;
 #endif
+}; // namespace log
```
namespace is called `logging`.

sfan5 commented on src/script/lua_api/l_env.cpp:
```diff
@@ -34,10 +34,11 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-				if (env == NULL) return 0
+#define GET_ENV_PTR ServerEnvironment* env = \
+		dynamic_cast<ServerEnvironment*>(getEnv(L)); \
+	if (env == NULL) return 0
```
This should be intended another tab.
Also this isn't related to logging in any way

sfan5 commented on src/script/lua_api/l_util.cpp:
```diff
@@ -36,17 +36,16 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+		 * differently than lua_tostring.
+		 */
+		lua_pushvalue(L, -1);  // Copy tostring
+		lua_pushvalue(L, i);   // Copy 'i'th parameter
```
Why change something with the comments here?

sfan5 commented on src/subgame.h:
```diff
@@ -51,10 +51,7 @@ struct SubgameSpec
-	{
-		return (id != "" && path != "");
-	}
+	bool isValid() const { return !id.empty() && !path.empty(); }
```
How very related to logging...

Zeno- replied:
Why is this now on one line?

Also, !(id.empty() || path.empty()) might be clearer

kwolekr commented:
I like the addition of the warning stream, fixing Lua print(), but that's just about it.

After taking a closer look at the log.cpp refactor, I'm not sure it actually adds any value.  I mean we have functions transformed into methods of a C++ class, but how does this *make the code better*?  Like, just as a *single* example:  The addition of LogOutputEntry.  This is fundamentally a wrapper class around ILogOutput that separates out the silenced boolean.  How does this added abstraction make anything cleaner?  You're adding unnecessary shit that shouldn't really be there.  And then the usage of operator overloading!  I thought we agreed to not use operator overloading unless it would provide huge benefit otherwise.

kwolekr commented:
From a functional standpoint, this patch makes dstream become a standard logging level that cannot be disabled.

This goes against the whole point of dstream though!  dstream is a minimal, debugging log facility detached from the standard logging system so that it works even in the case that standard logging has some sort of problem.

ShadowNinja commented:
@kwolekr
  * Moving things into classes reduces the number of globals and keeps things organized.
  * I explained the need for LogOutputEntry in IRC, but I later thought of an even better way to implement it and did that.
  * How would you prefer I do the comparison?  A loop comparing it to the values in a static const array and returning the index?
  * If you need a more low-level API you should use `std::cerr` or similar.  I added `g_log_outputs_stream` which functions like the old `dstream` though.

est31 commented:
Also, we need dstream for unittests because there, the standard logging facilities are disabled, so that the test statistics printed using dstream are nicer shown in stdout.

est31 commented:
what does `FacePositionCache` have to do with logging? you should do that move inside a separate PR, and minetest names header files with .h, not .hpp.

rubenwardy commented:
> minetest names header files with .h, not .hpp..

Imo, this should be phased out. We use cpp not c.

est31 commented:
Well, its mostly a question of preference. Mine leans towards .h.

nerzhul commented:
We use only .h and this is our norm

rubenwardy commented:
Yes, I'm aware that we currently use .h, but it doesn't have to be like
that. While CPP is (mostly/kind of) backwards compatible with C, it's a
different language. Why should it share the same file extension?

ShadowNinja commented:
@est31: I was adding a util/numeric.h include and I didn't want to pull that in when it's only used in two (?) files.  You're right though, it shouldn't be in this PR, I'll separate it.

ShadowNinja commented:
@est31: I split the face_position_cache change out, but left it on this branch for now.  I changed the `.hpp` to `.h` (we can discuss a switch later).

est31 commented on src/util/numeric.h:
```diff
@@ -144,36 +125,17 @@ inline bool isInArea(v3s16 p, v3s16 d)
-
 inline v3s16 arealim(v3s16 p, s16 d)
 {
-	if(p.X < 0)
```
From the [kernel code style](https://www.kernel.org/doc/Documentation/CodingStyle)
```
Don't put multiple statements on a single line unless you have
something to hide:

	if (condition) do_this;
	  do_something_everytime;
```

ShadowNinja replied:
It helps if you put comments below the changed code so I don't have to search for it in the "Files changed" tab.  Also, there's a clear difference between that example and this: The following line isn't deceptively indented here.  But if that's what you want it can be done...

est31 replied:
In this case its even OK I think, because it appears in a long chain. Our code styles shouldn't stand in the way if something is obviously better. You can leave it I think.

est31 commented on src/log.h:
```diff
@@ -20,83 +20,166 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+namespace logging {
+
+/* Use this for logging everything.
+ * If you need to explicitly print something, use cout or cerr.
```
Perhaps revise this, as we have `rawstream` now?

ShadowNinja replied:
`rawstream` is part of the logging setup, and this implies that the list is of non-logging utilities (dstream used to be separate from the logging facilities in `log.h`).

est31 replied:
I see, ok then.

est31 commented on src/log.cpp:
```diff
@@ -19,195 +19,261 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+}
 
-	log_threadnames.erase(id);
+const std::string Logger::getLevelName(LogLevel lev)
```
Why is this inside the class, and `stringToLevel` isn't?

ShadowNinja replied:
Good question, I'm not sure...

ShadowNinja replied:
Found the answer:
`getLevelName` returns a label for display when printing.  It should probably be named `getLevelLabel` since it doesn't actually convert to the level's name (eg, `LL_NONE` becomes `""`).  It's in the Logger class since it's tied to the actual logging and isn't useful outside of the logger.

`stringToLevel`, on the other hand, is a utility that's used to convert a level name to a level (eg, `"none"` -> `LL_NONE` and `""` -> error/`LL_MAX`).

est31 replied:
For my server terminal, I'll need `getLevelLabel`. Currently it only prints the level number as integer.

ShadowNinja replied:
Huh?  You mean it prints `y-m-d h:m:s: 3[Thread]:` instead of `y-m-d h:m:s: ACTION[Thread]:`?

est31 replied:
![screenshot](https://cloud.githubusercontent.com/assets/8872119/8992732/733e4acc-3700-11e5-917f-d29e14c1f26d.png)
https://github.com/est31/minetest/tree/ncurses-console

ShadowNinja replied:
That's an issue with your code then, not this PR, right?

est31 replied:
Yes, my code, I'm waiting for the method to get exposed, so that I can use it.

est31 commented on src/main.cpp:
```diff
@@ -533,38 +508,48 @@ static bool read_config_file(const Settings &cmd_args)
+		case 3: conf_loglev = "info"; break;
+		case 4: conf_loglev = "verbose"; break;
+		default:
+			warningstream << "Supplied invalid debug_log_level!" << std::endl;
```
Perhaps use a sane default, instead of returning?

Zeno- replied:
Strange style. I don't think the body of the case should be on the same line. Even if that was acceptable then default should follow the same style. I also agree that case 0: return is pretty yuck (it's almost hidden).

ShadowNinja replied:
@est31: It defaults to "action" now.
@Zeno-: I restructured it, better?

Zeno- commented on src/util/numeric.h:
```diff
@@ -144,36 +125,17 @@ inline bool isInArea(v3s16 p, v3s16 d)
-		p.Y = d-1;
-	if(p.Z > d-1)
-		p.Z = d-1;
+	if (p.X < 0) p.X = 0;
```
Why are these on one line? (@est31 I don't think this should be ignored at all; it's bad style).

Also, unless (d - 1) can equal -1 then these should be more optimally arranged.

ShadowNinja replied:
Split the lines.  Also, what do you mean by more optimally?

nerzhul commented:
Jenkins build failed

nerzhul commented:
Jenkins build failed

nerzhul commented:
Jenkins build passed

ShadowNinja commented:
@nerzhul: It seems you have a bot posting build status in comments, please disable that, I can see the status in the place GitHub provides for it.

kwolekr commented:
This commit is absurd.
Please, NEVER do this again.  This is taking way too long to review because it's so large and unwieldy, and it's low priority but needs to be committed very fast due to the sheer number of dependencies.

A refactor of something is welcome but _please_ don't confuse a large number of line changes with being productive.

rubenwardy commented:
I don't get why you did s/__FUNCTION_NAME/FUNCTION_NAME/ that doesn't belong in this commit.

kwolekr commented:
Lots of things don't belong in this commit.

I'm splitting it into 6 separate ones:
- the actual log refactor
- one for the warningstream changes
- __FUNCTION_NAME change
- serverlist 'dedicated' announcement
- Lua deprecation warning modification
- Android startup modification

kwolekr commented:
Another few things:
- Stop squishing code blocks onto a single line.
- Stop changing empty string comparisons to use empty().
- Stop changing header.h includes to cheader.

These are annoying and DO NOT HELP MINETEST.

kwolekr commented:
https://github.com/minetest/minetest/commit/2139d7d45fb1a8ed250ad96c9975c581f02f72a9
https://github.com/minetest/minetest/commit/96cc5b34fe0151d2a1498eaaafb205757db346c4
https://github.com/minetest/minetest/commit/4236792b87a68819c56f3b6f8bd44f48058dec57
https://github.com/minetest/minetest/commit/6f4d6cb5741d10e10dacc78e264e421bf6a297cf
https://github.com/minetest/minetest/commit/6f2d785d0ff761961912d7e79a7d16c4adf86861
https://github.com/minetest/minetest/commit/659922fd30d5ee3d7876b22e4d26b116d1ae2513
https://github.com/minetest/minetest/commit/7b8d372947aae232ddf598155e972bb4dda157af
https://github.com/minetest/minetest/commit/e067ceacb8787f1dc5b1ab69b50dba4497c55433
